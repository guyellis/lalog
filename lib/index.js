const isObject = require('lodash/isObject');
const debug = require('debug');
const uuid = require('uuid');

const {
  logBatch,
  logSingle,
} = require('./loggly-wrapper');

/**
 * @type {Array<import('../typings').LevelEnum>}
 */
const levels = ['trace', 'info', 'warn', 'error', 'fatal', 'security'];
const errorLevel = levels.indexOf('error');

const getInitialLogLevel = () => {
  /** @type {import('../typings').LevelEnum} */
  // @ts-ignore Type 'string' is not assignable to type 'LevelEnum'
  const laLogLevel = process.env.LALOG_LEVEL;
  if (levels.includes(laLogLevel)) {
    return levels.indexOf(laLogLevel);
  }
  return levels.indexOf('error');
};

let currentLevelIndex = getInitialLogLevel();

/**
 * @type {import('../typings')}
 */
class Logger {
  /**
   * Create an instance of Logger
   * @param {import('../typings').LogOptions} options
   */
  constructor(options) {
    const {
      addTrackId,
      moduleName,
      presets,
      serviceName,
      isTransient,
    } = options;

    this.isTransient = !!isTransient;

    /**
     * @type {Array|null}
     */
    this.logCollector = isTransient ? [] : null;

    this.presets = {
      module: moduleName,
      ...(isObject(presets) ? presets : {}),
    };

    // @ts-ignore
    if (addTrackId && !this.presets.trackId) {
    // @ts-ignore
      this.presets.trackId = uuid.v4();
    }

    this.debug = debug(`${serviceName}:${moduleName}`);
    this.tag = `${serviceName}-${process.env.NODE_ENV}`;

    // Setup timeEnd so that it can be called without a level and when that happens the
    // level will default to info:
    const defaultTimeEndLevel = levels.indexOf('info');
    this.timeEnd = this.writeTimeEnd.bind(this, defaultTimeEndLevel);

    // Listed like this so that Typescript can type each log level.
    // Previously this was setup using a loop but Typescript couldn't type it
    this.trace = this.write.bind(this, levels.indexOf('trace'));
    this.info = this.write.bind(this, levels.indexOf('info'));
    this.warn = this.write.bind(this, levels.indexOf('warn'));
    this.error = this.write.bind(this, levels.indexOf('error'));
    this.fatal = this.write.bind(this, levels.indexOf('fatal'));
    this.security = this.write.bind(this, levels.indexOf('security'));

    // @ts-ignore
    this.timeEnd.trace = this.writeTimeEnd.bind(this, levels.indexOf('trace'));
    // @ts-ignore
    this.timeEnd.info = this.writeTimeEnd.bind(this, levels.indexOf('info'));
    // @ts-ignore
    this.timeEnd.warn = this.writeTimeEnd.bind(this, levels.indexOf('warn'));
    // @ts-ignore
    this.timeEnd.error = this.writeTimeEnd.bind(this, levels.indexOf('error'));
    // @ts-ignore
    this.timeEnd.fatal = this.writeTimeEnd.bind(this, levels.indexOf('fatal'));
    // @ts-ignore
    this.timeEnd.security = this.writeTimeEnd.bind(this, levels.indexOf('security'));

    this.timers = {};
    /**
     * Start a timer log - same as console.time()
     * @param {string} label - label to use when calling timeEnd()
     */
    this.time = (label) => {
    // @ts-ignore
      this.timers[label] = Date.now();
    };
  }

  /**
    * Create an instance of Logger
    * @static
    * @param {import('../typings').LogOptions} options
    * @returns {import('../typings')}
    * @memberof Logger
    */
  static create(options) {
    // @ts-ignore Type 'Logger' is not assignable to type 'LaLog'.
    // Property 'create' is missing in type 'Logger'.
    return new Logger(options);
  }

  /**
   * Get an array of all available log levels
   * @static
   * @returns {Array<import('../typings').LevelEnum>}
   * @memberof Logger
   */
  static allLevels() {
    return levels;
  }

  /**
   * Get the current log level
   * @static
   * @returns {import('../typings').LevelEnum}
   * @memberof Logger
   */
  static getLevel() {
    return levels[currentLevelIndex];
  }

  /**
   * Change the minimum level to write logs
   * @static
   * @param {import('../typings').LevelEnum} newLevelName
   * @returns {import('../typings').LevelEnum}
   * @memberof Logger
   */
  static setLevel(newLevelName) {
    const previousLevel = Logger.getLevel();
    const newLevelIndex = levels.indexOf(newLevelName);
    if (newLevelIndex >= 0) {
      currentLevelIndex = newLevelIndex;
    }
    return previousLevel;
  }

  /**
   * Parse the Express request (req) object for logging
   * @static
   * @param {Object} req
   * @returns {Object}
   * @memberof Logger
   */
  static parseReq(req) {
    return {
      body: req.body,
      headers: req.headers,
      method: req.method,
      params: req.params,
      path: req.path,
      query: req.query,
      url: req.url,
      user: req.user,
    };
  }

  /**
   * Format milliseconds to a string for logging
   * @static
   * @param {number} milliseconds
   * @returns {string}
   * @memberof Logger
   */
  static formatMilliseconds(milliseconds) {
    const date = new Date(0);
    date.setMilliseconds(milliseconds);
    return date.toISOString().substr(11, 12);
  }

  /**
   * [Private] Write the timer label end
   * @param {number} levelIndex
   * @param {string} label
   * @param {object} [extraLogData={}]
   * @returns {Promise}
   * @memberof Logger
   */
  writeTimeEnd(levelIndex, label, extraLogData = {}) {
    // @ts-ignore
    const time = this.timers[label];
    const duration = Object.prototype.hasOwnProperty.call(this.timers, label)
      ? Logger.formatMilliseconds(Date.now() - time)
      : `Missing label "${label}" in timeEnd()\n${(new Error()).stack}`;

    let { msg } = extraLogData;
    msg = msg ? `Timer - ${msg}` : 'Timer';

    const logData = {
      ...extraLogData,
      msg,
      timerLabel: label,
      duration,
    };
    return this.write(levelIndex, logData);
  }

  /**
   * Write log to destination
   * @param {number} levelIndex
   * @param {object} logData
   * @param {object=} response
   * @returns {Promise}
   * @memberof Logger
   */
  async write(levelIndex, logData, response) {
    if (!isObject(logData)) {
      // eslint-disable-next-line no-console
      console.error(`Expecting an object in logger write method but got "${typeof logData}"`);
      return Promise.resolve();
    }

    /**
     * @type {object}
     */
    const logObj = { ...this.presets, ...logData };

    if (response) {
      // If the response object has been included with the call then it means we need to
      // send an error message with an errorId.
      // Prima facie this seems like a terrible idea but it seems to work well as
      // a response can be immediately sent to the client which includes a code
      // that can be provided to the user and links back to the error log.
      const errorId = logObj.errorId || uuid.v4();
      logObj.errorId = errorId;
      const { res, code } = response;
      res.status(code).send({ success: false, errorId });
    }

    // When do we log?
    // - If !isTransient and levelIndex >= currentLevelIndex
    //   - normal logging - current logic
    // - If isTransient and levelIndex < currentLevelIndex
    //   - push this log item onto the array
    // - If isTransient and levelIndex >= currentLevelIndex and !isTransientTriggered
    //   - set isTransientTriggered to true
    //   - push this log item onto the array
    //   - bulk log everything in array
    //   - empty array (for early GC)
    // - If isTransientTriggered
    //   - log everything


    if (levelIndex >= currentLevelIndex || this.isTransient) {
      logObj.level = levels[levelIndex];

      if (levelIndex >= errorLevel && !logObj.err) {
        logObj.err = new Error();
      }

      if (logObj.err) {
        if (!logObj.err.stack) {
          // This will happen if we manually created an err prop - it might not have a stack prop
          logObj.err.stack = new Error().stack;
        }
        logObj.fullStack = logObj.err.stack.split('\n').slice(1);
        /**
         * Checks if string includes node_modules
         * @param {string} i
         */
        const hasNodeModules = (i) => !i.includes('/node_modules/');
        logObj.shortStack = logObj.fullStack.filter(hasNodeModules);
        if (!logObj.msg) {
          logObj.msg = logObj.err.message;
        }
        delete logObj.err;
      }

      if (logObj.req) {
        logObj.req = Logger.parseReq(logObj.req);
      }

      if (this.logCollector !== null && !this.isTransientTriggered) {
        this.logCollector.push(logObj);
        if (levelIndex >= currentLevelIndex) {
          // Need to batch log here
          this.isTransientTriggered = true;
          await logBatch({ tag: this.tag, logObj: this.logCollector });
          this.logCollector = null; // Can GC right away now that this array is no longer needed
        }
      } else {
        this.debug(logObj);
        return logSingle({ tag: this.tag, logObj });
      }
    }
    return Promise.resolve();
  }
}

module.exports = Logger;
