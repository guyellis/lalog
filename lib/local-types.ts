import { Request, Response } from 'express';

export type BetterOmit<T, K extends PropertyKey> =
  { [P in keyof T as Exclude<P, K>]: T[P] };

export const levels = ['trace', 'info', 'warn', 'error', 'fatal', 'security'] as const;
export type LevelType = typeof levels[number];

export interface LogPresets extends Record<string, unknown> {
    module?: string;
    trackId?: string;
  }

export interface LaLogOptions {
    addTrackId?: boolean;
    moduleName?: string;
    presets?: LogPresets;
    serviceName?: string;
    isTransient?: boolean;
  }

export type ParseReqIn = Request & { user?: unknown };
export type ParseReqOut = Pick<ParseReqIn, 'body' |
  'headers' |
  'method' |
  'params' |
  'path' |
  'query' |
  'url' |
  'user'>;

export interface LogData extends Record<string, unknown> {
    err?: Error;
    msg?: string;
    req?: ParseReqIn;
  }

export interface LogDataOut extends BetterOmit<LogData, 'req'>, LogPresets {
    /**
     * The Stack property from the Error object split into lines.
     */
    fullStack?: string[];
    /**
    * Created from the fullStack by removing lines containing node_modules
    */
    shortStack?: string[];
    req?: ParseReqOut;
  }

export interface ResponseWrapper {
    res: Response;
    code: number;
  }

export type LogFunction = (logData: LogData, response?: ResponseWrapper) => Promise<any>;
export type TimeLogFunction = (
    label: string, level: LevelType, extraLogDat?: LogData,
  ) => Promise<any>;

export type logDataEnriched = Omit<LogDataOut, 'err'> & {
    err: Record<string, unknown>
  }
