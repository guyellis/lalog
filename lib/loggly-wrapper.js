const _ = require('lodash');
const fetch = require('node-fetch');

/**
 * Write log to Loggly
 * @param {object} options
 * @param {string} options.tag
 * @param {string=} options.logglyToken
 * @param {string} options.logObj
 * @param {boolean} bulk
 */
const log = async (options, bulk) => {
  const {
    tag,
    logglyToken = process.env.LOGGLY_TOKEN,
    logObj: body,
  } = options;


  const pathPart = bulk ? 'bulk' : 'inputs';

  if (logglyToken) {
    /**
     * @type {string}
     */
    let url = '';
    /**
     * @type {import('node-fetch').RequestInit}
     */
    let fetchOptions = {};
    try {
      url = `https://logs-01.loggly.com/${pathPart}/${logglyToken}/tag/${tag}/`;
      fetchOptions = {
        headers: {
          'Content-Type': 'application/json',
          Accept: 'application/json',
        },
        method: 'POST',
        body,
      };

      /**
       * @type {import('node-fetch').Response}
       */
      // @ts-ignore Cannot invoke an expression whose type lacks a call signature...
      // ts-ignore seems to be needed because of a bug in the node-fetch typing file.
      // Try and remove the ignore if this typing file is updated
      const result = await fetch(url, fetchOptions);

      if (result.status !== 200) {
        const { stack } = new Error();
        // eslint-disable-next-line no-console
        console.error(`fetch() call failed with ${result.status}\n${stack}`);
        return {};
      }

      return result.json();
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error(`fetch() threw an error: ${err.message}
url: ${url}
options: ${JSON.stringify(fetchOptions, null, 2)}`);
      return {};
    }
  }
  // eslint-disable-next-line no-console
  console.warn('loggly token has not been defined');

  return {};
};

/**
 * Log a single log object
 * @param {object} options
 * @param {string} options.tag
 * @param {string=} options.logglyToken
 * @param {object} options.logObj
 */
const logSingle = (options) => {
  const { logObj } = options;
  if (!_.isObject(logObj)) {
    // eslint-disable-next-line no-console
    console.error(`Expected an Object in logSingle but got ${typeof logObj}`);
    return Promise.resolve();
  }

  const body = JSON.stringify(logObj);
  return log({
    ...options,
    logObj: body,
  }, false);
};

/**
 * Log an array of logs
 * @param {object} options
 * @param {string} options.tag
 * @param {string=} options.logglyToken
 * @param {Array<object>} options.logObj
 */
const logBatch = async (options) => {
  const { logObj } = options;
  if (!_.isArray(logObj)) {
    // eslint-disable-next-line no-console
    console.error(`Expected an Array in logBatch but got ${typeof logObj}`);
    return Promise.resolve();
  }

  // https://stackoverflow.com/questions/52248377/typescript-array-map-with-json-stringify-produces-error
  // @ts-ignore - no idea why typescript can't handle this .map()
  const body = logObj.map(JSON.stringify).join('\n');
  return log({
    ...options,
    logObj: body,
  }, true);
};

module.exports = {
  logBatch,
  logSingle,
};
