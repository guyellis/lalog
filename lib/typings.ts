// Type definitions for lalog 0.6.1
// Project: https://github.com/guyellis/lalog
// Definitions by: Guy Ellis <https://github.com/guyellis>

/* ~ This declaration specifies that the class constructor function
 *~ is the exported object from the file
 */

/* ~ Write your module's methods and properties in this class */

type LevelEnum =
  'trace'|
  'info'|
  'warn'|
  'error'|
  'fatal'|
  'security';

interface LogPresets {
  [key: string]: string;
}

interface LaLogOptions {
  addTrackId?: boolean;
  moduleName?: string;
  presets?: LogPresets;
  serviceName?: string;
  isTransient?: boolean;
}

interface TimeEndLog {
  (label: string, extraLogDat?: any): Promise<any>;
  trace?: (label: string, extraLogDat?: any) => Promise<any>;
  info?: (label: string, extraLogDat?: any) => Promise<any>;
  warn?: (label: string, extraLogDat?: any) => Promise<any>;
  error?: (label: string, extraLogDat?: any) => Promise<any>;
  fatal?: (label: string, extraLogDat?: any) => Promise<any>;
  security?: (label: string, extraLogDat?: any) => Promise<any>;
}

interface LaLog{
  new(options: LaLogOptions): void;

  // Static methods
  create: (options: LaLogOptions) => LaLog;
  setLevel: (level: LevelEnum) => LevelEnum;
  getLevel: () => LevelEnum;
  allLevels: () => Array<LevelEnum>;

  // Instance methods
  time: (label: string) => void;
  // TODO: timeEnd has all the log methods on it.
  timeEnd: (label: string, extraLogData?: object) => Promise<undefined>;
  trace: (logObj: object) => Promise<undefined>;
  info: (logObj: object) => Promise<undefined>;
  warn: (logObj: object) => Promise<undefined>;
  error: (logObj: object) => Promise<undefined>;
  fatal: (logObj: object) => Promise<undefined>;
  security: (logObj: object) => Promise<undefined>;
}
