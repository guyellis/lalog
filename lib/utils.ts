export const isObject = (
  obj?: object | null,
): boolean => !!obj && obj.toString() === '[object Object]' && !Array.isArray(obj);

type CircularReplacer = (key: any, value: any) => any;

// From MDN:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value
const getCircularReplacer = (): CircularReplacer => {
  const seen = new WeakSet();
  return (key, value): any => {
    if (typeof value === 'object' && value !== null) {
      if (seen.has(value)) {
        return '[CIRCULAR]';
      }
      seen.add(value);
    }
    return value;
  };
};

export const safeJsonStringify = (obj: any): string => JSON.stringify(obj, getCircularReplacer());
