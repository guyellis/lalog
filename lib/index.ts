import { v4 } from 'uuid';
import { Request, Response } from 'express';
import { isObject } from './utils';

import {
  logBatch,
  logSingle,
} from './loggly-wrapper';

const levels = ['trace', 'info', 'warn', 'error', 'fatal', 'security'] as const;
export type LevelType = typeof levels[number];

export interface LogPresets {
  [key: string]: string | undefined;
  module?: string;
  trackId?: string;
}

export interface LaLogOptions {
  addTrackId?: boolean;
  moduleName?: string;
  presets?: LogPresets;
  serviceName?: string;
  isTransient?: boolean;
}

export type LogFunction = (logData: any, response?: any) => Promise<any>;
export type TimeLogFunction = (label: string, level?: LevelType, extraLogDat?: any) => Promise<any>;

const errorLevel = levels.indexOf('error');

const getInitialLogLevel = (): number => {
  const laLogLevel = process.env.LALOG_LEVEL as LevelType;
  if (levels.includes(laLogLevel || '')) {
    return levels.indexOf(laLogLevel);
  }
  return levels.indexOf('error');
};

let currentLevelIndex = getInitialLogLevel();

export type ParseReqInOut = Request & { user?: unknown };

interface LogData extends Record<string, unknown> {
  msg?: string;
}

interface ResponseWrapper {
  res: Response;
  code: number;
}

export default class Logger {
  isTransient: boolean;

  logCollector: any[] | null;

  presets: LogPresets;

  tag: string;

  timeEnd: TimeLogFunction;

  trace: LogFunction;

  info: LogFunction;

  warn: LogFunction;

  error: LogFunction;

  fatal: LogFunction;

  security: LogFunction;

  time: (label: string) => void;

  timers: Record<string, number>;

  isTransientTriggered?: boolean;

  constructor(options: LaLogOptions) {
    const {
      addTrackId,
      moduleName,
      presets,
      serviceName,
      isTransient,
    } = options;

    this.isTransient = !!isTransient;

    this.logCollector = isTransient ? [] : null;

    this.presets = {
      module: moduleName,
      ...(isObject(presets) ? presets : {}),
    };

    if (addTrackId && !this.presets.trackId) {
      this.presets.trackId = v4();
    }

    this.tag = `${serviceName}-${process.env.NODE_ENV}`;

    this.timeEnd = this.writeTimeEnd.bind(this);

    // Listed like this so that Typescript can type each log level.
    // Previously this was setup using a loop but Typescript couldn't type it
    this.trace = this.write.bind(this, levels.indexOf('trace'));
    this.info = this.write.bind(this, levels.indexOf('info'));
    this.warn = this.write.bind(this, levels.indexOf('warn'));
    this.error = this.write.bind(this, levels.indexOf('error'));
    this.fatal = this.write.bind(this, levels.indexOf('fatal'));
    this.security = this.write.bind(this, levels.indexOf('security'));

    this.timers = {};
    /**
     * Start a timer log - same as console.time()
     * @param label - label to use when calling timeEnd()
     */
    this.time = (label: string): void => {
      this.timers[label] = Date.now();
    };
  }

  /**
    * Create an instance of Logger
    */
  static create(options: LaLogOptions): Logger {
    return new Logger(options);
  }

  /**
   * Get an array of all available log levels
   */
  static allLevels(): ReadonlyArray<LevelType> {
    return levels;
  }

  /**
   * Get the current log level
   */
  static getLevel(): LevelType {
    return levels[currentLevelIndex];
  }

  /**
   * Change the minimum level to write logs.
   * @param newLevelName - If falsy it will just return the current level
   * @returns The previous level
   */
  static setLevel(newLevelName?: LevelType): LevelType {
    const previousLevel = Logger.getLevel();
    if (!newLevelName) {
      return previousLevel;
    }
    const newLevelIndex = levels.indexOf(newLevelName);
    if (newLevelIndex >= 0) {
      currentLevelIndex = newLevelIndex;
    }
    return previousLevel;
  }

  /**
   * Parse the Express request (req) object for logging
   */
  static parseReq(req: ParseReqInOut): Partial<ParseReqInOut> {
    return {
      body: req.body,
      headers: req.headers,
      method: req.method,
      params: req.params,
      path: req.path,
      query: req.query,
      url: req.url,
      user: req.user,
    };
  }

  /**
   * Format milliseconds to a string for logging
   */
  static formatMilliseconds(milliseconds: number): string {
    const date = new Date(0);
    date.setMilliseconds(milliseconds);
    return date.toISOString().substr(11, 12);
  }

  /**
   * Write the timer label end
   */
  writeTimeEnd(label: string, level?: LevelType, extraLogDat?: LogData): Promise<any> {
    const levelIndex = levels.indexOf(level ?? 'info');
    const extraLogData = extraLogDat || {};
    const time = this.timers[label];
    const duration = Object.prototype.hasOwnProperty.call(this.timers, label)
      ? Logger.formatMilliseconds(Date.now() - time)
      : `Missing label "${label}" in timeEnd()\n${(new Error()).stack}`;

    let { msg } = extraLogData;
    msg = msg ? `Timer - ${msg}` : 'Timer';

    const logData = {
      ...extraLogData,
      duration,
      msg,
      timerLabel: label,
    };
    return this.write(levelIndex, logData);
  }

  /**
   * Write log to destination
   */
  async write(levelIndex: number, logData: LogData, response?: ResponseWrapper): Promise<any> {
    if (!isObject(logData)) {
      // eslint-disable-next-line no-console
      console.error(`Expecting an object in logger write method but got "${typeof logData}"`);
      return Promise.resolve();
    }

    const logObj: any = { ...this.presets, ...logData };

    if (response) {
      // If the response object has been included with the call then it means we need to
      // send an error message with an errorId.
      // Prima facie this seems like a terrible idea but it seems to work well as
      // a response can be immediately sent to the client which includes a code
      // that can be provided to the user and links back to the error log.
      const errorId = logObj.errorId || v4();
      logObj.errorId = errorId;
      const { res, code } = response;
      res.status(code).send({ errorId, success: false });
    }

    // When do we log?
    // - If !isTransient and levelIndex >= currentLevelIndex
    //   - normal logging - current logic
    // - If isTransient and levelIndex < currentLevelIndex
    //   - push this log item onto the array
    // - If isTransient and levelIndex >= currentLevelIndex and !isTransientTriggered
    //   - set isTransientTriggered to true
    //   - push this log item onto the array
    //   - bulk log everything in array
    //   - empty array (for early GC)
    // - If isTransientTriggered
    //   - log everything

    if (levelIndex >= currentLevelIndex || this.isTransient) {
      logObj.level = levels[levelIndex];

      if (levelIndex >= errorLevel && !logObj.err) {
        logObj.err = new Error();
      }

      if (logObj.err) {
        if (!logObj.err.stack) {
          // This will happen if we manually created an err prop - it might not have a stack prop
          logObj.err.stack = new Error().stack;
        }
        logObj.fullStack = logObj.err.stack.split('\n').slice(1);
        /**
         * Checks if string includes node_modules
         */
        const hasNodeModules = (i: string): boolean => !i.includes('/node_modules/');
        logObj.shortStack = logObj.fullStack.filter(hasNodeModules);
        if (!logObj.msg) {
          logObj.msg = logObj.err.message;
        }
        delete logObj.err;
      }

      if (logObj.req) {
        logObj.req = Logger.parseReq(logObj.req);
      }

      if (this.logCollector !== null && !this.isTransientTriggered) {
        this.logCollector.push(logObj);
        if (levelIndex >= currentLevelIndex) {
          // Need to batch log here
          this.isTransientTriggered = true;
          await logBatch({ logObj: this.logCollector, tag: this.tag });
          this.logCollector = null; // Can GC right away now that this array is no longer needed
        }
      } else {
        return logSingle({ logObj, tag: this.tag });
      }
    }
    return Promise.resolve();
  }
}
