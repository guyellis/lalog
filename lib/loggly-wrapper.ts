import fetch, { RequestInit, Response } from 'node-fetch';
import { isObject, safeJsonStringify } from './utils';

interface LogOptions {
  tag: string;
  logglyToken?: string;
  logObj: string;
}

const log = async (
  options: LogOptions, bulk: boolean,
): Promise<Record<string, unknown>> => {
  const {
    tag,
    logglyToken = process.env.LOGGLY_TOKEN,
    logObj: body,
  } = options;

  const pathPart = bulk ? 'bulk' : 'inputs';

  if (logglyToken) {
    let url = '';
    let fetchOptions: RequestInit = {};
    try {
      url = `https://logs-01.loggly.com/${pathPart}/${logglyToken}/tag/${tag}/`;
      fetchOptions = {
        body,
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        method: 'POST',
      };

      const result: Response = await fetch(url, fetchOptions);

      if (result.status !== 200) {
        const { stack } = new Error();
        // eslint-disable-next-line no-console
        console.error(`fetch() call failed with ${result.status}\n${stack}`);
        return {};
      }

      return result.json();
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error(`fetch() threw an error: ${err.message}
url: ${url}
options: ${safeJsonStringify(fetchOptions as Record<string, unknown>)}`);
      return {};
    }
  }
  // eslint-disable-next-line no-console
  console.warn('loggly token has not been defined');

  return {};
};

export interface LogSingleOptions {
  tag: string;
  logglyToken?: string;
  logObj: any;
}

export const logSingle = (
  options: LogSingleOptions,
): Promise<Record<string, unknown> | void> => {
  const { logObj } = options;
  if (!isObject(logObj)) {
    // eslint-disable-next-line no-console
    console.error(`Expected an Object in logSingle but got ${typeof logObj}`);
    return Promise.resolve();
  }

  const body = safeJsonStringify(logObj);
  return log({
    ...options,
    logObj: body,
  }, false);
};

export interface LogBatchOptions {
  tag: string;
  logglyToken?: string;
  logObj: any[];
}

export const logBatch = async (
  options: LogBatchOptions,
): Promise<Record<string, unknown> | void> => {
  const { logObj } = options;
  if (!Array.isArray(logObj)) {
    // eslint-disable-next-line no-console
    console.error(`Expected an Array in logBatch but got ${typeof logObj}`);
    return Promise.resolve();
  }

  const body = logObj.map(safeJsonStringify).join('\n');
  return log({
    ...options,
    logObj: body,
  }, true);
};
