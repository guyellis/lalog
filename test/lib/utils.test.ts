import { isObject, safeJsonStringify } from '../../lib/utils';

describe('utils', () => {
  test('isObject', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    const emptyArrow = (): void => {};

    expect(isObject({})).toBe(true);
    expect(isObject(1 as unknown as object)).toBe(false);
    expect(isObject(true as unknown as object)).toBe(false);
    expect(isObject(null)).toBe(false);
    expect(isObject(emptyArrow)).toBe(false);
    expect(isObject([1, 2])).toBe(false);
    expect(isObject([])).toBe(false);
  });

  test('safeJsonStringify', () => {
    const obj1 = {
      one: 'one',
      two: {
        three: 3,
      },
    };
    let result = safeJsonStringify(obj1);
    expect(result).toMatchSnapshot();

    result = safeJsonStringify(null);
    expect(result).toMatchSnapshot();

    result = safeJsonStringify('i am string');
    expect(result).toMatchSnapshot();
  });

  test('safeJsonStringify circular', () => {
    const circular: Record<string, unknown> = {
      one: 'one',
      two: {
        three: 3,
      },
    };
    circular.four = circular;
    const result = safeJsonStringify(circular);
    expect(result).toMatchSnapshot();
  });
});
