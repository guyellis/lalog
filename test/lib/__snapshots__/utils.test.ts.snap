// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`utils safeJsonStringify 1`] = `"{"one":"one","two":{"three":3}}"`;

exports[`utils safeJsonStringify 2`] = `"null"`;

exports[`utils safeJsonStringify 3`] = `""i am string""`;

exports[`utils safeJsonStringify circular 1`] = `"{"one":"one","two":{"three":3},"four":"[CIRCULAR]"}"`;
